src/haz3lcore/statics/TermBase.re

```diff
@@ -243,6 +254,104 @@ and Exp: {
     | Cast(t, Typ.t, Typ.t)
   and t = IdTagged.t(term);
 
+  let rec term_of_menhir_ast = (exp: Haz3lmenhir.AST.exp): Exp.term => {
+    switch (exp) {
+    | InvalidExp(s) => Invalid(s)
+    | Int(i) => Int(i)
+    | Float(f) => Float(f)
+    | String(s) => String(s)
+    | Bool(b) => Bool(b)
+    | Var(x) => Var(x)
+    | Constructor(x, ty) => Constructor(x, Typ.of_menhir_ast(ty))
+    | Deferral(pos) =>
+      switch (pos) {
+      | InAp => Deferral(InAp)
+      | OutsideAp => Deferral(OutsideAp)
+      }
+    | ListExp(l) => ListLit(List.map(of_menhir_ast, l))
+    | TupleExp(t) =>
+      if (List.length(t) == 1) {
+        Parens(of_menhir_ast(List.hd(t)));
+      } else {
+        Tuple(List.map(of_menhir_ast, t));
+      }
+    | Let(p, e1, e2) =>
+      Let(Pat.of_menhir_ast(p), of_menhir_ast(e1), of_menhir_ast(e2))
+    | FixF(p, e) => FixF(Pat.of_menhir_ast(p), of_menhir_ast(e), None)
+    | TypFun(t, e) => TypFun(TPat.of_menhir_ast(t), of_menhir_ast(e), None)
+    | Undefined => Undefined
+    | TyAlias(tp, ty, e) =>
+      TyAlias(
+        TPat.of_menhir_ast(tp),
+        Typ.of_menhir_ast(ty),
+        of_menhir_ast(e),
+      )
+    | BuiltinFun(s) => BuiltinFun(s)
+    | DeferredAp(f, a) => DeferredAp(of_menhir_ast(f), [of_menhir_ast(a)])
```

Comment:
Is there a reason the menhir ast only has a single deferral?

Reply:
Looks like this is true for a few other cases below. I'll have to read further to understand what's going on.

Reply:
Sorry, this may just have been an error on my part? What exactly should the deferral look like?

Reply:
Based on my current understanding it should be the list of function parameters and `Deferral(InAp)` for all of the deferrals present.

So for `f(_, 1, _)` it would be `(Deferral(InAp), Int(8), Deferral(InAp))`. It's probably worth synthesizing a few of these in the editor and seeing the DHExp/Exp structure to compare and add tests. I'll leave comments on the other cases with a singleton list.